clippy_lints/src/string_patterns.rs

```diff
@@ -0,0 +1,223 @@
+
+const PATTERN_METHODS: [(&str, usize); 22] = [
+    ("contains", 0),
+    ("starts_with", 0),
+    ("ends_with", 0),
+    ("find", 0),
+    ("rfind", 0),
+    ("split", 0),
+    ("split_inclusive", 0),
+    ("rsplit", 0),
+    ("split_terminator", 0),
+    ("rsplit_terminator", 0),
+    ("splitn", 1),
+    ("rsplitn", 1),
+    ("split_once", 0),
+    ("rsplit_once", 0),
+    ("matches", 0),
+    ("rmatches", 0),
+    ("match_indices", 0),
+    ("rmatch_indices", 0),
+    ("trim_start_matches", 0),
+    ("trim_end_matches", 0),
+    ("replace", 0),
+    ("replacen", 0),
+];
+
+fn check_single_char_pattern_lint(cx: &LateContext<'_>, arg: &Expr<'_>, applicability: &mut Applicability) {
+    if let Some(hint) = get_hint_if_single_char_arg(cx, arg, applicability, true) {
+        span_lint_and_sugg(
+            cx,
+            SINGLE_CHAR_PATTERN,
+            arg.span,
+            "single-character string constant used as pattern",
+            "consider using a `char`",
+            hint,
+            *applicability,
+        );
+    }
+}
+
+fn check_manual_pattern_char_comparison(cx: &LateContext<'_>, arg: &Expr<'_>, applicability: Applicability) {
+    if let ExprKind::Closure(closure) = arg.kind
+        && let body = cx.tcx.hir().body(closure.body)
+        && let Some(param_name) = body.params[0].pat.simple_ident()
+    {
+        let ex = body.value;
+        let mut set_chars: Vec<String> = Vec::new();
+
+        // We want to retrieve all the comparisons done.
+        // They are ordered in a nested way and so we need to traverse the AST to collect them all.
+        for_each_expr(ex, |sub_expr| -> ControlFlow<(), Descend> {
+            match sub_expr.kind {
+                ExprKind::Binary(op, left, right) => {
+                    // Avoid matching code that contains complex comparisons such as functions calls etc..
+                    // (ex: c.is_whitespace())
+                    if matches!(
+                        left.kind,
+                        ExprKind::Binary(_, _, _) | ExprKind::Match(_, _, _) | ExprKind::Lit(_) | ExprKind::Path(_)
+                    ) && matches!(
+                        right.kind,
+                        ExprKind::Binary(_, _, _) | ExprKind::Match(_, _, _) | ExprKind::Lit(_) | ExprKind::Path(_)
+                    ) {
+                        if matches!(op.node, BinOpKind::Eq | BinOpKind::Or) {
+                            ControlFlow::Continue(Descend::Yes)
+                        } else {
+                            ControlFlow::Continue(Descend::No)
+                        }
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
+                ExprKind::Match(match_value, arms, _) => {
+                    if snippet(cx, match_value.span, "") != param_name.name.as_str() || arms.len() != 2 {
+                        return ControlFlow::Break(());
+                    }
```

Comment:
A nit, destructure `arms` here as `[arm, _]` rather than checking the len + indexing

We should check that `sub_expr` is an expansion of `matches!()` so that we don't let through other things. We can do that with `matching_root_macro_call(cx, sub_expr.span, sym::matches_macro).is_some()`. For example

```rust
"".find(|c| match c {
    'a' | 'b' => true,
    _ => c.is_ascii(),
});
```

We also need to check that `arm` has no guard:

```rust
"".find(|c| matches!(c, 'a' | 'b' if false));
```

Reply:
Applied. I learnt a lot of helper ! :o